mem_db.go

```diff
@@ -157,113 +183,144 @@ func (db *MemDB) NewBatch() Batch {
 
 // Implements DB.
 func (db *MemDB) Iterator(start, end []byte) (Iterator, error) {
-	db.mtx.Lock()
-	defer db.mtx.Unlock()
+	db.mtx.RLock()
+	defer db.mtx.RUnlock()
 
-	keys := db.getSortedKeys(start, end, false)
-	return newMemDBIterator(db, keys, start, end), nil
+	return newMemDBIterator(db.btree, start, end, false), nil
 }
 
 // Implements DB.
 func (db *MemDB) ReverseIterator(start, end []byte) (Iterator, error) {
-	db.mtx.Lock()
-	defer db.mtx.Unlock()
+	db.mtx.RLock()
+	defer db.mtx.RUnlock()
 
-	keys := db.getSortedKeys(start, end, true)
-	return newMemDBIterator(db, keys, start, end), nil
+	return newMemDBIterator(db.btree, start, end, true), nil
 }
 
-// We need a copy of all of the keys.
-// Not the best, but probably not a bottleneck depending.
 type memDBIterator struct {
-	db    DB
-	cur   int
-	keys  []string
-	start []byte
-	end   []byte
+	ch     <-chan *item
+	cancel context.CancelFunc
+	item   *item
+	start  []byte
+	end    []byte
 }
 
 var _ Iterator = (*memDBIterator)(nil)
 
-// Keys is expected to be in reverse order for reverse iterators.
-func newMemDBIterator(db DB, keys []string, start, end []byte) *memDBIterator {
-	return &memDBIterator{
-		db:    db,
-		cur:   0,
-		keys:  keys,
-		start: start,
-		end:   end,
+func newMemDBIterator(bt *btree.BTree, start []byte, end []byte, reverse bool) *memDBIterator {
+	ctx, cancel := context.WithCancel(context.Background())
+	ch := make(chan *item, 64)
```

Comment:
why 64? maximum 64 items?

Reply:
We buffer 64 items in the channel. This allows the goroutines to do more work per context switch. If the channel is unbuffered, then you get 2 context switches per item, which kills performance. Benchmarks showed 64 to be the sweet spot.

I'll add constants for these values.